README.md
# pino-slack-webhook

🌲 Send message to Slack via webhook

## Installation

```console
npm i pino-slack-webhook
```

## Usage

```console
$ pino-slack-webhook --help
pino-slack-webhook [options]

Sending
  -m, --method
             [string] [choices: "POST", "PUT", "PATCH", "GET"] [default: "POST"]
      --url      url to send logs to                         [string] [required]
  -t, --timeout  timeout (in ms) to send logs in bucket that are not filled
                                                        [number] [default: 5000]

Basic Auth
  -u, --username  basic auth username                                   [string]
  -p, --password  basic auth password                                   [string]

Retry
  -r, --retries   number of retries to do if failure       [number] [default: 5]
  -i, --interval  interval (in ms) to retry sending if failure
                                                        [number] [default: 1000]

Slack
      --unfurlLinks  Enables or disables link unfurling
                                                      [boolean] [default: false]
      --unfurlMedia  Enables or disables media unfurling
                                                      [boolean] [default: false]
  -m, --mrkdwn       Enables or disables mrkdwn formatting
                                                       [string] [default: false]

Options:
      --help     Show help                                             [boolean]
      --version  Show version number                                   [boolean]
  -l, --log      log to console as well               [boolean] [default: false]
      --silent   silence pino-slack-webhook logs for failures and retries
                                                      [boolean] [default: false]
  -c, --config   path to json config                                    [string]
```

## Environment Variables

All options can be defined in the environment and are prefixed with `PINO_SLACK_WEBHOOK_`. All
camel-cased options are parsed with delimiters of `_`.

_e.g. The option `unfurlLinks` as an env var would be `PINO_SLACK_WEBHOOK_UNFURL_LINKS`._

## URL

**Example**

```console
node . | pino-slack-webhook --url=https://hooks.slack.com/services/xxx/xxx/xxx
```

## Body Type

- `ndjson` - New-line delimited JSON. See [ndjson](https://github.com/ndjson/ndjson-spec)
- `json` - Standard JSON sending of data. Logs are sent in the format of

  ```json
  {
    "logs": [...logs]
  }
  ```

## Auth

Currently only basic auth is implemented for the CLI usage. For header usage, you can see the API usage.

## API

You can also use this module as a [pino destination](https://github.com/pinojs/pino/blob/master/docs/api.md#destination).

This will use the same batching function like the CLI usage. If the batch length
is not reached within a certain time (`timeout`), it will auto "flush".

### `createWriteStream`

The options passed to this follow the same values as the CLI defined above.

| Property    | Type                    | Required/Default |
| ----------- | ----------------------- | ---------------- |
| url         | `string`                | REQUIRED         |
| log         | `boolean`               | false            |
| silent      | `boolean`               | false            |
| method      | `string`                | "POST"           |
| username    | `string`                |                  |
| password    | `string`                |                  |
| headers     | `Record<string,string>` |                  |
| retries     | `number`                | 5                |
| interval    | `number`                | 1000             |
| timeout     | `number`                | 5000             |
| unfurlLinks | `boolean`               | false            |
| unfurlMedia | `boolean`               | false            |
| mrkdwn      | `boolean`               | false            |
| config      | `string`                |                  |

```ts
import { createWriteStream } from 'pino-slack-webhook';

const stream = createWriteStream({
  url: 'https://hooks.slack.com/services/xxx/xxx/xxx',
});

const logger = pino(
  {
    level: 'info',
  },
  stream,
);

logger.info('test log!');
```


CHANGELOG.md
## [1.2.1](https://github.com/shrujalshah28/pino-slack-webhook/compare/v1.2.0...v1.2.1) (2020-12-31)


### Bug Fixes

* make some payload property optional ([bb884c3](https://github.com/shrujalshah28/pino-slack-webhook/commit/bb884c314275178124478ac5858b7672d2259c0f))

# [1.2.0](https://github.com/shrujalshah28/pino-slack-webhook/compare/v1.1.1...v1.2.0) (2020-12-31)


### Features

* allow to format log message ([765bf00](https://github.com/shrujalshah28/pino-slack-webhook/commit/765bf00da33428b0536e39a0e88b33d3baf5f4d5))

## [1.1.1](https://github.com/shrujalshah28/pino-slack-webhook/compare/v1.1.0...v1.1.1) (2020-12-31)


### Performance Improvements

* remove source map ([4a75cf6](https://github.com/shrujalshah28/pino-slack-webhook/commit/4a75cf64597ed1f8be1e6d7de977d33a6b8f642f))

# [1.1.0](https://github.com/shrujalshah28/pino-slack-webhook/compare/v1.0.1...v1.1.0) (2020-12-31)


### Features

* add slack specific options ([f87127a](https://github.com/shrujalshah28/pino-slack-webhook/commit/f87127ab819111f0d52fb2989f52414e36bd786c))

## [1.0.1](https://github.com/shrujalshah28/pino-slack-webhook/compare/v1.0.0...v1.0.1) (2020-12-26)

### Performance Improvements

- remove batchSize argument ([05c2efb](https://github.com/shrujalshah28/pino-slack-webhook/commit/05c2efb0fd39dc13b4eeaeca30253793f69a78f8))

# 1.0.0 (2020-12-25)

### Features

- setup project ([0628f81](https://github.com/shrujalshah28/pino-slack-webhook/commit/0628f81aed0b01704dfeb0418d28ac4e7f32811e))


src/send.ts
import got, { Method } from 'got';

import { args, SlackPayload } from './args';
import { logError, logWarn } from './log';

export type Body = {
  json: SlackPayload;
};

export function createBody(log: Record<string, unknown>): Body {
  const { unfurlLinks = false, unfurlMedia = false, mrkdwn = false, format } = args;

  const payload: SlackPayload = {
    unfurl_links: unfurlLinks,
    unfurl_media: unfurlMedia,
    mrkdwn: mrkdwn,
    text: `${log.level}: ${log.msg}`,
  };

  if (format && typeof format === 'function') {
    const layout = format(log);

    // Note: Supplying `text` when `blocks` is also supplied will cause `text`
    // to be used as a fallback for clients/surfaces that don't suopport blocks
    payload.text = layout.text ?? undefined;
    payload.attachments = layout.attachments ?? undefined;
    payload.blocks = layout.blocks ?? undefined;
  }

  return { json: payload };
}

export function send(log: Record<string, unknown>, numRetries = 0): void {
  const {
    url,
    method,
    username,
    password,
    headers = {},
    retries = 5,
    interval = 1000,
    silent = false,
  } = args;

  const limitHit = numRetries === retries;

  // fire and forget so we don't await or anything
  got(url, {
    method: method as Method,
    username,
    password,
    headers,
    allowGetBody: true,
    ...createBody(log),
  })
    .then()
    .catch(err => {
      if (!silent) {
        logError(err, limitHit ? null : `...retrying in ${interval}ms`);
      }

      if (limitHit) {
        if (!silent) {
          // make sure to stringify to get the whole thing, e.g. don't want
          // cutoffs on deep objects...
          logWarn(`max retries hit (${retries}). dropping logs:`, JSON.stringify(log));
        }

        return;
      }

      numRetries++;

      setTimeout(() => send(log, numRetries), interval);
    });
}


src/args.ts
import { readFileSync } from 'fs';
import yargs from 'yargs';

export type SlackPayload = {
  // eslint-disable-next-line camelcase
  unfurl_links?: boolean;
  // eslint-disable-next-line camelcase
  unfurl_media?: boolean;
  mrkdwn?: boolean;
  text: string;
  attachments?: Array<unknown>;
  blocks?: Array<Record<string, unknown>>;
};

export type Args = {
  url: string;
  unfurlLinks?: boolean;
  unfurlMedia?: boolean;
  mrkdwn?: boolean;
  log?: boolean;
  silent?: boolean;
  method?: string;
  username?: string;
  password?: string;
  headers?: Record<string, string>;
  retries?: number;
  interval?: number;
  timeout?: number;
  config?: string;
  format?: (log: Record<string, unknown>) => SlackPayload;
};

/**
 * Used with API and not with CLI...
 */
const defaultArgs: Args = {
  log: false,
  silent: true,
  method: 'POST',
  url: '',
  retries: 5,
  interval: 1000,
  timeout: 5000,
  unfurlLinks: false,
  unfurlMedia: false,
  mrkdwn: false,
};

export let args: Args;

/**
 * On demand loading of args since CLI would conflict with the loading of these.
 */
export function loadArgs(): Args {
  args = yargs
    .usage('pino-slack-webhook [options]')
    .env('PINO_SLACK_WEBHOOK')
    .option('log', {
      alias: 'l',
      type: 'boolean',
      desc: 'log to console as well',
      default: false,
    })
    .option('silent', {
      type: 'boolean',
      desc: 'silence pino-slack-webhook logs for failures and retries',
      default: false,
    })
    .option('method', {
      alias: 'm',
      type: 'string',
      choices: ['POST', 'PUT', 'PATCH', 'GET'],
      group: 'Sending',
      default: 'POST',
    })
    .option('url', {
      type: 'string',
      require: true,
      group: 'Sending',
      desc: 'url to send logs to',
    })
    .option('username', {
      alias: 'u',
      type: 'string',
      desc: 'basic auth username',
      group: 'Basic Auth',
    })
    .option('password', {
      alias: 'p',
      type: 'string',
      desc: 'basic auth password',
      group: 'Basic Auth',
    })
    .option('retries', {
      alias: 'r',
      type: 'number',
      desc: 'number of retries to do if failure',
      group: 'Retry',
      default: 5,
    })
    .option('interval', {
      alias: 'i',
      type: 'number',
      desc: 'interval (in ms) to retry sending if failure',
      group: 'Retry',
      default: 1000,
    })
    .option('timeout', {
      alias: 't',
      type: 'number',
      desc: 'timeout (in ms) to send logs in bucket that are not filled',
      group: 'Sending',
      default: 5000,
    })
    .option('unfurlLinks', {
      type: 'boolean',
      desc: 'Enables or disables link unfurling',
      group: 'Slack',
      default: false,
    })
    .option('unfurlMedia', {
      type: 'boolean',
      desc: 'Enables or disables media unfurling',
      group: 'Slack',
      default: false,
    })
    .option('mrkdwn', {
      alias: 'm',
      type: 'boolean',
      desc: 'Enables or disables mrkdwn formatting',
      group: 'Slack',
      default: false,
    })
    .option('config', {
      alias: 'c',
      type: 'string',
      desc: 'path to json config',
      config: true,
      configParser: path => JSON.parse(readFileSync(path, 'utf8')),
    })
    .parse();

  return args;
}

/**
 * Sets args to passed in values defaulting to above defaults.
 * Used in CLI ONLY.
 * @param newArgs
 */
export function setArgs(newArgs: Partial<Args>): void {
  args = {
    ...defaultArgs,
    ...newArgs,
  };
}
